pkg/logger/logger_test.go

```diff
@@ -0,0 +1,246 @@
+	h.wg.Add(1)
+	defer h.wg.Done()
+	h.handler.ServeHTTP(w, r)
+}
+
+// simple handler that logs the request
+// and sleeps for 5 seconds
+// to simulate a long running request
+func makeLoggingHandler(ctx context.Context) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		logger := logging.FromContext(ctx)
+		select {
+		case <-time.After(5 * time.Second):
+			// simulate a long running request, in case ctx.Done() (srv.ShutdownServer)
+			// is not called fast enough
+		case <-ctx.Done():
+			break
+			// the context will be canceled when the server is shutting down
+			// this gives an exact timing of workload simulation
+		}
+		logger.Infof("Received request:%s %s", r.Method, r.URL) // use our logger to test concurrency
+	}
+}
+
+func syslogServer(port string, messages chan<- string, ready chan<- struct{}, shutdownSyslog <-chan struct{}) {
+	// start listening on the port
+	ln, err := net.Listen("tcp", ":"+port)
+	if err != nil {
+		log.Fatalf("Error starting the server: %v", err)
+	}
+
+	// Notify that server is ready
+	ready <- struct{}{}
+
+	var wg sync.WaitGroup
+	var connections []net.Conn
+	var connectionsLock sync.Mutex
+
+	// Shutdown logic
+	go func() {
+		<-shutdownSyslog
+		ln.Close()
+		// closing connections to force the scanner to stop
+		connectionsLock.Lock()
+		defer connectionsLock.Unlock()
+		for _, conn := range connections {
+			conn.Close()
+		}
+	}()
+
+	// start accepting connections
+	for {
+		conn, err := ln.Accept()
+		if err != nil {
+			if opErr, ok := err.(*net.OpError); ok && opErr.Op == "accept" {
+				// Listener closed, stopping accepting connections
+				break
+			}
+			log.Printf("Error accepting connection: %v", err)
+			continue
+		}
+
+		// new connection, add it to the list
+		connectionsLock.Lock()
+		connections = append(connections, conn)
+		connectionsLock.Unlock()
+
+		// read from the connection
+		// and send the messages to the message channel
+		wg.Add(1)
+		go func(conn net.Conn) {
+			defer wg.Done()
+			defer conn.Close()
+
+			scanner := bufio.NewScanner(conn)
+			for scanner.Scan() {
+				text := scanner.Text()
+				messages <- text
+			}
+		}(conn)
+	}
+
+	// no more incoming connections, wait for all connections to be closed
+	wg.Wait()
+	close(messages)
+}
+
+func startHTTPServer(
+	ctx context.Context,
+	port string,
+	handler http.Handler,
+	ready chan<- struct{},
+	shutdownHTTP <-chan struct{},
+	shutdownSyslog chan<- struct{},
+) {
+	var wg sync.WaitGroup
+	customHandler := NewCustomHandler(&wg, handler)
+	server := &http.Server{
+		Addr:    ":" + port,
+		Handler: customHandler,
+	}
+
+	// Start the server
+	go func() {
+		if err := server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			log.Fatalf("Error starting the server: %v", err)
+		}
+	}()
+
+	// Wait until the server is ready
+	timeout := time.After(5 * time.Second)           // Wait up to 5 seconds for server to start
```

Comment:
http server should be ready within `1s`. something is wrong if it takes longer than this.